src/java.base/share/classes/sun/security/ssl/ServerHello.java

```diff
@@ -794,6 +794,15 @@ public byte[] produce(ConnectionContext context,
             hhrm.write(shc.handshakeOutput);
             shc.handshakeOutput.flush();
 
+            // In the Middlebox Compatibility Mode the server sends a
+            // dummy change_cipher_spec record immediately after its
+            // first handshake message. This may either be after
+            // a ServerHello or a HelloRetryRequest.
+            //(RFC 8446, Appendix D.4)
+            shc.conContext.outputRecord.changeWriteCiphers(
```

Comment:
the JDK exposes middlebox compatibility mode via the `jdk.tls.client.useCompatibilityMode` property. Is that a factor for this fix ?

Reply:
No, in that case, clientHello.sessionId.length() will be zero, and it won't send CCS.